cpp/include/cuvs/neighbors/ivf_flat.hpp

```diff
@@ -0,0 +1,331 @@
+  [[nodiscard]] constexpr inline auto metric() const noexcept -> cuvs::distance::DistanceType
+  {
+    return static_cast<cuvs::distance::DistanceType>((int)raft_index_->metric());
+  }
+  /** Whether `centers()` change upon extending the index (ivf_pq::extend). */
+  [[nodiscard]] constexpr inline auto adaptive_centers() const noexcept -> bool
+  {
+    return raft_index_->adaptive_centers();
+  }
+  /**
+   * Inverted list data [size, dim].
+   *
+   * The data consists of the dataset rows, grouped by their labels (into clusters/lists).
+   * Within each list (cluster), the data is grouped into blocks of `kIndexGroupSize` interleaved
+   * vectors. Note, the total index length is slightly larger than the source dataset length,
+   * because each cluster is padded by `kIndexGroupSize` elements.
+   *
+   * Interleaving pattern:
+   * within groups of `kIndexGroupSize` rows, the data is interleaved with the block size equal to
+   * `veclen * sizeof(T)`. That is, a chunk of `veclen` consecutive components of one row is
+   * followed by a chunk of the same size of the next row, and so on.
+   *
+   * __Example__: veclen = 2, dim = 6, kIndexGroupSize = 32, list_size = 31
+   *
+   *     x[ 0, 0], x[ 0, 1], x[ 1, 0], x[ 1, 1], ... x[14, 0], x[14, 1], x[15, 0], x[15, 1],
+   *     x[16, 0], x[16, 1], x[17, 0], x[17, 1], ... x[30, 0], x[30, 1],    -    ,    -    ,
+   *     x[ 0, 2], x[ 0, 3], x[ 1, 2], x[ 1, 3], ... x[14, 2], x[14, 3], x[15, 2], x[15, 3],
+   *     x[16, 2], x[16, 3], x[17, 2], x[17, 3], ... x[30, 2], x[30, 3],    -    ,    -    ,
+   *     x[ 0, 4], x[ 0, 5], x[ 1, 4], x[ 1, 5], ... x[14, 4], x[14, 5], x[15, 4], x[15, 5],
+   *     x[16, 4], x[16, 5], x[17, 4], x[17, 5], ... x[30, 4], x[30, 5],    -    ,    -    ,
+   *
+   */
+  /** Sizes of the lists (clusters) [n_lists]
+   * NB: This may differ from the actual list size if the shared lists have been extended by another
+   * index
+   */
+  inline auto list_sizes() noexcept -> raft::device_vector_view<uint32_t, uint32_t>
+  {
+    return raft_index_->list_sizes();
+  }
+  [[nodiscard]] inline auto list_sizes() const noexcept
+    -> raft::device_vector_view<const uint32_t, uint32_t>
+  {
+    return raft_index_->list_sizes();
+  }
+
+  /** k-means cluster centers corresponding to the lists [n_lists, dim] */
+  inline auto centers() noexcept -> raft::device_matrix_view<float, uint32_t, raft::row_major>
+  {
+    return raft_index_->centers();
+  }
+  [[nodiscard]] inline auto centers() const noexcept
+    -> raft::device_matrix_view<const float, uint32_t, raft::row_major>
+  {
+    return raft_index_->centers();
+  }
+
+  /**
+   * (Optional) Precomputed norms of the `centers` w.r.t. the chosen distance metric [n_lists].
+   *
+   * NB: this may be empty if the index is empty or if the metric does not require the center norms
+   * calculation.
+   */
+  inline auto center_norms() noexcept -> std::optional<raft::device_vector_view<float, uint32_t>>
```

Comment:
Shall we have always int64_t as matrix index type?
```suggestion
  inline auto center_norms() noexcept -> std::optional<raft::device_vector_view<float, int64_t>>
```

Reply:
Same here, valid concern, but this is how the RAFT API is defined at the moment. I think we could update this in a follow-up PR.